Add tests for distributions moments

michaelosthege opened the pull request:
Just checking the CI already.

Do we want to systematically test the moments?
Any ideas how to do that without lots of copy-paste?

Related to #5078

ricardoV94 commented:
> Do we want to systematically test the moments?

Definitely

> Any ideas how to do that without lots of copy-paste?

None yet. We should probably test that it respects the RV size, while also working with broadcasted params.

ricardoV94 commented on pymc/distributions/continuous.py:
```diff
@@ -789,6 +780,20 @@ def logp(
 
+    def get_moment(value_var, size, mu, sigma, lower, upper):
+        initval = None
+        has_lower = at.isinf(lower).eval()
```
This can probably work with a nested switch that does not require knowing lower/upper at call time.

Otherwise you can check if it's a constant and retrieve the value. If it's not a constant the `eval` will fail with missing input

ricardoV94 commented on pymc/tests/test_distributions_moments.py:
```diff
@@ -0,0 +1,140 @@
+    assert_moment_is_expected(model, expected)
+
+
+@pytest.mark.skip(reason="aeppl interval transform fails when both edges are None")
```
https://github.com/aesara-devs/aeppl/pull/82

michaelosthege commented on pymc/distributions/distribution.py:
```diff
@@ -371,7 +370,7 @@ def get_moment(rv: TensorVariable) -> TensorVariable:
     """
     size = rv.owner.inputs[1]
-    return _get_moment(rv.owner.op, rv, size, *rv.owner.inputs[3:])
+    return _get_moment(rv.owner.op, rv, size, *rv.owner.inputs[3:]).astype(aesara.config.floatX)
```
Good placement!

ricardoV94 replied:
Actually just realized it should convert to the RVs dtype which maybe be int64

michaelosthege commented on pymc/distributions/shape_utils.py:
```diff
@@ -674,3 +675,8 @@ def maybe_resize(
     return rv_out
+
+
+def rv_size_is_none(size: Variable) -> bool:
+    """Check wether an rv size is None (ie., at.Constant([]))"""
+    return isinstance(size, Constant) and size.data.size == 0
```
It's perfectly fine to do `pm.get_moment(..., size=())`, right? Or even with `size=[]`?
Maybe this utility function should account for `size` being `()`, `[]` or `None` too?

ricardoV94 replied:
No, the size comes from the RV and is always converted to an empty string when it's None IIRC

michaelosthege reviewed (commented):
I can't approve it, even though @ricardoV94 did 98 % of the changes.

ricardoV94 committed:
Add tests for distribution moments

Fixes some pre-existing moments
Adds moments for HalfNormal and TruncatedNormal distributions
Adds helper rv_size_is_none function

twiecki commented:
Do we still need more moments?

ricardoV94 commented:
> Do we still need more moments?

Yes, see #5078

patel-zeel commented on pymc/distributions/continuous.py:
```diff
@@ -828,6 +850,12 @@ def dist(cls, sigma=None, tau=None, sd=None, *args, **kwargs):
         return super().dist([0.0, sigma], **kwargs)
 
+    def get_moment(rv, size, loc, sigma):
+        moment = loc + sigma
```
I have a doubt here.

Current implementation: loc + sigma
[Wikipedia version](https://en.wikipedia.org/wiki/Half-normal_distribution): loc + \sqrt{\frac{2}{\pi}}sigma

ricardoV94 replied:
I think you are right, we should probably be using that. Do you want to open a PR to fix it?

patel-zeel replied:
Sure, happily :)

ricardoV94 replied:
A bit more involved, we should probably also use the truncated normal mean as the moment. We might even be able to simplify the switch statement, if the normal logcdf behaves well with +- infinity values: https://en.wikipedia.org/wiki/Truncated_normal_distribution
